model/bitmap-constructor.js

```diff
@@ -0,0 +1,160 @@
+};
+Bitmap.prototype.setColorArray = function(colors) {
+  var dibSize = this.buf.readUInt32LE(14);
+  var numColors = this.buf.readInt32LE(46);
+  //TODO: Check length of colors = this.getColorArray()
+  if(colors.length !== numColors) {
+    // What does in here? Return an error?
+    // return new Error;
+    throw new Error('number of colors does not match up');
+  }
+
+  for (let i = 0; i < numColors; i++) {
+    // if (colors[i])
+    if (typeof(colors[i]) !== 'object') {
+      // return error? what do we want to do here?
+      // return new Error;
+      throw new Error('colors array does not contain objects');
+    }
+    if ((!colors[i].hasOwnProperty('blue')) || (!colors[i].hasOwnProperty('red')) || (!colors[i].hasOwnProperty('green')) || (!colors[i].hasOwnProperty('alpha'))) {
+      // return error? what do we want to do here?
+      // return new Error;
+      throw new Error('color object does not contain correct keys');
+      // )
+    }
+    if ((colors[i].blue > 255) || (colors[i].red > 255) || (colors[i].green > 255) || (colors[i].alpha > 255)) {
+      throw new Error('color values are out of bounds');
+    }
+    if ((colors[i].blue < 0) || (colors[i].red < 0) || (colors[i].green < 0) || (colors[i].alpha < 0)) {
+      throw new Error('color values are negative');
+    }
+  }
+
+  var position = dibSize + 14;
+  for (let i = 0; i < numColors; i++) {
+    this.buf.writeUInt8(colors[i].blue, position++);
+    this.buf.writeUInt8(colors[i].green, position++);
+    this.buf.writeUInt8(colors[i].red, position++);
+    this.buf.writeUInt8(colors[i].alpha, position++);
+  }
+  return this; //To chain seters
+};
+Bitmap.prototype.setPixelArray = function(pixels) {
+  if(!Array.isArray(pixels)) {
+    throw new Error('invalid pixel array');
+  }
+  //tODO: Possibly make a getPixelOffset method
+  var pixelOffset = this.buf.readUInt32LE(10);
+  var numPixels = this.getWidth() * this.getHeight();
+
+  if(pixels.length > numPixels) {
+    throw new Error('param pixels is too long to fit in this bitmap');
+  }
+
+  // pixels should be an array of integers in the range [0, numColors -1];
+  var numColors = this.getNumColors();
+  var numToWrite = Math.min(pixels.length, numPixels);
+  pixels = pixels.slice(0,numToWrite);
+  pixels.forEach(function(pixel) {
+    if(!Number.isInteger(pixel)) {
+      throw new Error('invalid color value');
+    }
+    if(pixel < 0 || pixel > numColors - 1) {
+      throw new Error('color value out of bounds');
+    }
+  });
+
+  for(let i = 0; i < numToWrite; i++) {
+    var offset = pixelOffset + i;
+    this.buf.writeUInt8(pixels[i], offset);
+  }
+
+  return this; //To chain seters
+};
+
+Bitmap.prototype.transform = function(operator) {
+  if(!operator) throw new Error('missing operator');
+  //TODO: Determine if operator is a function that takes in a bitmap?!?
+  operator(this); //Let the operator mess with this
+  return this; //To chain transforms
+};
+
+module.exports = exports = {};
```

Comment:
@dbecker4130 @geoffsimons @meganreardon  no need for this, you should only be exporting the constructor, it's prototype methods will be available after instantiation

ex: `module.exports = Bitmap` and then instantiate as needed in any other file